From 2d6b7590486f84d0517401c37c70d7eed95e715d Mon Sep 17 00:00:00 2001
From: Matthew Flatt <mflatt@racket-lang.org>
Date: Mon, 24 Nov 2025 17:54:09 -0700
Subject: [PATCH] arm32: repair for certain large/unaligned floating-point
 dereferences (#997)

When an offset to a floating-point field or array element is unaligned
or large enough and does not fit into a 12-bit encoding used for
immediate offsets, compilation could fail.
---
 mats/ftype.ms                    | 13 +++++++++++++
 release_notes/release_notes.stex | 10 ++++++++--
 s/arm32.ss                       |  7 ++++++-
 3 files changed, 27 insertions(+), 3 deletions(-)

diff --git a/mats/ftype.ms b/mats/ftype.ms
index ec1ea841..35c94e63 100644
--- a/mats/ftype.ms
+++ b/mats/ftype.ms
@@ -1397,6 +1397,19 @@
    (define-syntax unsigned-16 (make-compile-time-value "Non-interfering binding"))
    (let ([fptr (make-ftype-pointer unsigned-16 0)])
      (= (ftype-pointer-address fptr) 0)))
+
+  ;; regression test for arm32 backend and distant offset to floating-point field
+  (let ()
+    (define-ftype T
+      (struct
+          [a (array 6368 int)] ; offset derived from 6368 is not funky12
+          [n double]))
+    (define (f x y)
+      (ftype-set! T (n) x y))
+    (let ([p (make-ftype-pointer T (foreign-alloc (ftype-sizeof T)))])
+      (f p 1.0)
+      (equal? (ftype-ref T (n) p) 1.0)))
+
 )
 
 (mat ftype-pointer-address-optimizations
diff --git a/release_notes/release_notes.stex b/release_notes/release_notes.stex
index cda9ee59..b1a22d65 100644
--- a/release_notes/release_notes.stex
+++ b/release_notes/release_notes.stex
@@ -161,10 +161,10 @@ foreign functions may be compiled to a more efficient form, particularly if the
 accepts or returns flonums that may be unboxed (in the sense of
 Section~\ref{sec:unbox-floats}).
 
-\subsection{Improved flonum unboxing for \scheme{fptr-ref} and \scheme{fptr-set!} (10.3.0)}
+\subsection{Improved flonum unboxing for \scheme{ftype-ref} and \scheme{ftype-set!} (10.3.0)}
 
 Unboxing optimizations (in the sense of Section~\ref{sec:unbox-floats}) are more
-consistently applied for \scheme{fptr-ref} and \scheme{fptr-set!} on \scheme{double} and
+consistently applied for \scheme{ftype-ref} and \scheme{ftype-set!} on \scheme{double} and
 \scheme{float} fields.
 
 \subsection{Atomic compare-and-set for pairs (10.3.0)}
@@ -2903,6 +2903,12 @@ When \scheme{>} and \scheme{>=} are applied to two non-real arguments, the first
 is now reported as incorrect instead of the second. In general, when two or more arguments
 are invalid, there is no guarantee about which one is reported invalid.
 
+\subsection{Fix 32-bit ARM floating-point dereference offset (10.4.0)}
+
+A bug that affects compilation on 32-bit ARM systems has been fixed. The bug was triggered
+when an offset to a floating-point field or array element was unaligned or too large to
+fit into the 12-bit encoding used for immediate offsets.
+
 \subsection{Exit with failure on error from boot file (10.3.0)}
 
 When an error is encountered within a boot file---either raised
diff --git a/s/arm32.ss b/s/arm32.ss
index 44dae1ba..fe809c00 100644
--- a/s/arm32.ss
+++ b/s/arm32.ss
@@ -245,7 +245,12 @@
                     ;; offset not aligned or out of range
                     (let ([u (make-tmp 'umov)])
                       (seq
-                       (build-set! ,u (asm ,null-info ,(asm-add #f) ,x0 (immediate ,imm)))
+                       (if (funky12 imm)
+                           (build-set! ,u (asm ,null-info ,(asm-add #f) ,x0 (immediate ,imm)))
+                           (let ([tmp (make-tmp 'tmp)])
+                             (seq
+                              (build-set! ,tmp (immediate ,imm))
+                              (build-set! ,u (asm ,null-info ,(asm-add #f) ,x0 ,tmp)))))
                        (if (eq? x1 %zero)
                            (return u %zero 0)
                            (seq
